openquake/risklib/riskinput.py

```diff
@@ -0,0 +1,382 @@
+        self.weight = 0
+        for assets in self.assets_by_site:
+            for asset in assets:
+                taxonomies.add(asset.taxonomy)
+            self.weight += len(assets)
+        self.taxonomies = sorted(taxonomies)
+        self.tags = None  # for API compatibility with RiskInputFromRuptures
+        self.epsilons = {}  # populated with the function set_epsilons
+
+    @property
+    def imt_taxonomies(self):
+        """Return a list of pairs (imt, taxonomies) with a single element"""
+        return [(self.imt, self.taxonomies)]
+
+    def get_all(self):
+        """
+        Return dictionaries with
+        assets, hazards and epsilons for each taxonomy.
+        """
+        assets, hazards, epsilons = [], [], []
+        for hazard, assets_ in zip(self.hazard_by_site, self.assets_by_site):
+            for asset in assets_:
+                assets.append(asset)
+                hazards.append({self.imt: hazard})
+                epsilons.append(self.epsilons.get(asset.id, None))
+        return assets, hazards, epsilons
+
+    def __repr__(self):
+        return '<%s IMT=%s, taxonomy=%s, weight=%d>' % (
+            self.__class__.__name__, self.imt, ', '.join(self.taxonomies),
+            self.weight)
+
+
+def make_eps_dict(assets_by_site, num_samples, seed, correlation):
+    """
+    :param riskinput: an object with an attribute .assets_by_site
+    :param int num_samples: the number of ruptures
+    :param int seed: a random seed
+    :param float correlation: the correlation coefficient
+    :returns: dictionary asset_id -> epsilons
+    """
+    eps_dict = {}  # asset_id -> epsilons
+    all_assets = (a for assets in assets_by_site for a in assets)
+    assets_by_taxo = groupby(all_assets, operator.attrgetter('taxonomy'))
+    for taxonomy, assets in assets_by_taxo.iteritems():
+        shape = (len(assets), num_samples)
+        logging.info('Building %s epsilons for taxonomy %s', shape, taxonomy)
+        zeros = numpy.zeros(shape)
+        epsilons = scientific.make_epsilons(zeros, seed, correlation)
+        for asset, eps in zip(assets, epsilons):
+            eps_dict[asset.id] = eps
+    return eps_dict
+
+
+def expand(array, N, indices=None):
+    """
+    Given a non-empty array with n elements, expands it to a larger
+    array with N elements.
+
+    >>> expand([1], 3)
+    array([1, 1, 1])
+    >>> expand([1, 2, 3], 10)
+    array([1, 2, 3, 1, 2, 3, 1, 2, 3, 1])
+    >>> expand(numpy.zeros((2, 10)), 5).shape
+    (5, 10)
+    >>> expand([1, 2], 2)  # already expanded
+    [1, 2]
+    """
+    n = len(array)
+    if n == 0:
+        raise ValueError('Empty array')
+    elif n >= N:
+        return array
+    return numpy.array([array[i % n] for i in indices or xrange(N)])
+
+
+class RiskInputFromRuptures(object):
+    """
+    Contains all the assets associated to the given IMT and a subsets of
+    the ruptures for a given calculation.
+
+    :param imt_taxonomies: list given by the risk model
```

Comment:
Please check if the right convention is to describe the params here or inside the `__init__`